libinfinity/common/inf-file-util.c

```diff
@@ -518,47 +518,62 @@ inf_file_util_delete(const gchar* path,
 }
 
 /**
- * inf_file_write_data_with_perms:
+ * inf_file_util_write_private_data:
  * @filename: Filename of the file to be written to.
  * @data: Data that should be written to file.
  * @length: Length of data in bytes.
- * @flags: File access mode flags of the file to be open()'ed.
- * @mode: File permission bits of the file to be open()'ed.
  * @error: Location to store error information, if any, or %NULL.
  *
- * Writes @data to the file pointed to by @filename. Files may be created with
- * specified permissions (@mode). If the function fails, %FALSE is returned
- * and @error is set.
+ * Writes @data to the file pointed to by @filename. @filename is created with
+ * 0600 permission, if the file does not exist. If the function fails, %FALSE
+ * is returned and @error is set.
  *
  * Returns: %TRUE on success or %FALSE on error.
  */
 gboolean
-inf_file_write_data_with_perms(const gchar* filename,
-                               const void* data,
-                               size_t length,
-                               int flags,
-                               int mode,
-                               GError** error)
+inf_file_util_write_private_data(const gchar* filename,
+                                 const void* data,
+                                 size_t length,
+                                 GError** error)
 {
-  int fd = g_open(filename, flags, mode);
+  gchar *temp_file = g_strconcat(filename, ".XXXXXX", NULL);
+  gint fd = g_mkstemp_full(
+    temp_file,
+    O_WRONLY|O_CREAT|O_TRUNC,
+    0600
+  );
   if (fd == -1)
   {
     inf_file_util_set_error_from_errno(error, errno);
+    g_free(temp_file);
     return FALSE;
   }
-  ssize_t remaining = length;
+  size_t remaining = length;
   while (remaining > 0)
   {
     ssize_t written = write(fd, data, remaining);
-    if (remaining == -1)
+    if (written == -1)
     {
-      g_close(fd, error);
       inf_file_util_set_error_from_errno(error, errno);
+      g_close(fd, NULL);
+      g_free(temp_file);
       return FALSE;
     }
     remaining -= written;
```

Comment:
`data` needs to be advanced by `written` bytes at this point, or in the next iteration of the loop it will point to an incorrect location.